task/google/resources/resource_instance_template.go

```diff
@@ -0,0 +1,226 @@
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+
+	"terraform-provider-iterative/task/google/client"
+	"terraform-provider-iterative/task/universal"
+)
+
+func NewInstanceTemplate(client *client.Client, identifier string, defaultNetwork *DefaultNetwork, firewallRules []*FirewallRule, image *Image, credentials *Credentials, task universal.Task) *InstanceTemplate {
+	i := new(InstanceTemplate)
+	i.Client = client
+	i.Identifier = universal.NormalizeIdentifier(identifier, true)
+	i.Attributes = task
+	i.Dependencies.Credentials = credentials
+	i.Dependencies.DefaultNetwork = defaultNetwork
+	i.Dependencies.FirewallRules = firewallRules
+	i.Dependencies.Image = image
+	return i
+}
+
+type InstanceTemplate struct {
+	Client       *client.Client
+	Identifier   string
+	Attributes   universal.Task
+	Dependencies struct {
+		*DefaultNetwork
+		FirewallRules []*FirewallRule
+		*Image
+		*Credentials
+	}
+	Resource *compute.InstanceTemplate
+}
+
+func (i *InstanceTemplate) Create(ctx context.Context) error {
+	keyPair, err := i.Client.GetKeyPair(ctx)
+	if err != nil {
+		return err
+	}
+
+	publicKey, err := keyPair.PublicString()
+	if err != nil {
+		return err
+	}
+
+	formattedPublicKey := fmt.Sprintf("%s:%s host\n", i.Dependencies.Image.Attributes.SSHUser, strings.TrimSpace(publicKey))
+
+	if i.Attributes.Environment.Variables == nil {
+		i.Attributes.Environment.Variables = make(map[string]*string)
+	}
+	for name, value := range *i.Dependencies.Credentials.Resource {
+		valueCopy := value
+		i.Attributes.Environment.Variables[name] = &valueCopy
+	}
+
+	script := i.Attributes.Environment.GenerateMachineScript()
+
+	size := i.Attributes.Size.Machine
+	sizes := map[string]string{
+		"m":       "e2-custom-8-32768",
+		"l":       "e2-custom-32-131072",
+		"xl":      "n2-custom-64-262144",
+		"m+k80":   "custom-8-53248+nvidia-tesla-k80*1",
+		"l+k80":   "custom-32-131072+nvidia-tesla-k80*4",
+		"xl+k80":  "custom-64-212992-ext+nvidia-tesla-k80*8",
+		"m+v100":  "custom-8-65536-ext+nvidia-tesla-v100*1",
+		"l+v100":  "custom-32-262144-ext+nvidia-tesla-v100*4",
+		"xl+v100": "custom-64-524288-ext+nvidia-tesla-v100*8",
+	}
+
+	if val, ok := sizes[size]; ok {
+		size = val
+	}
+
+	accelerators := []*compute.AcceleratorConfig{}
+
+	match := regexp.MustCompile(`^([^+]+)(?:\+([^*]+)\*([1-9]\d*))?$`).FindStringSubmatch(size)
+	if match == nil {
+		return errors.New("invalid machine type")
+	}
+
+	machineType := match[1]
+	if match[2] != "" {
+		acceleratorType := match[2]
+		acceleratorCount, err := strconv.Atoi(match[3])
+		if err != nil {
+			return err
+		}
+		accelerators = append(accelerators, &compute.AcceleratorConfig{
+			AcceleratorCount: int64(acceleratorCount),
+			AcceleratorType:  acceleratorType,
+		})
+	}
+
+	var isPreemptible bool
+	switch i.Attributes.Spot {
```

Comment:
```go
if i.Attributes.Spot > 1 {
  return errors.New("preemptible instances don't have bidding price")
}
isPreemptible := i.Attributes.Spot != 0
```

Reply:
`switch` kill it before it grows!

Reply:
> _`if i.Attributes.Spot > 1 {`_
Did you mean `> 0` instead?

Reply:
Your suggestion is perhaps a bit too lax with the accepted values. Not that I complain, but this `switch` statement is probably nicer than allowing values like `-1.25` or `0.33` for disabling the spot mode. 😈

Reply:
My code is totally equivalent to that horrendous switch being more elegant handling the error first 
🎩

Reply:
> _totally equivalent_

`spot = -0.141592` 😈

Reply:
Still, I like it

Reply:
> spot = -0.141592

thats silly, that would mean that aws would pay you to use them. I can not imagine someone doing that... but its doable

Reply:
> -1.25 or 0.33 for disabling the spot mode. 😈

only negative numbers but I can fix it

Reply:
My point is that any negative value would disable preemptive mode in Google Cloud. Still, it's fine.

Reply:
It's fine: any negative number can mean "disabled": it's probably better than enforcing `-1`

Reply:
fixed ☝️